persistence/src/main/java/cz/muni/fi/pa165/sportsclub/dao/CoachDaoImpl.java

```diff
@@ -0,0 +1,64 @@
+package cz.muni.fi.pa165.sportsclub.dao;
+
+import cz.muni.fi.pa165.sportsclub.entity.Coach;
+import org.springframework.stereotype.Repository;
+import org.springframework.transaction.annotation.Transactional;
+
+import javax.persistence.EntityManager;
+import javax.persistence.PersistenceContext;
+import java.util.List;
+
+/**
+ * Implementation of CoachDAO
+ * @author Jan Cech
+ */
+@Repository
+@Transactional
+public class CoachDaoImpl implements CoachDao {
+
+    @PersistenceContext
+    private EntityManager entityManager;
+
+    @Override
+    public void create(Coach coach) {
+        if (coach == null) {
+            throw new IllegalArgumentException("Coach cannot be null");
+        }
+        entityManager.persist(coach);
+    }
+
+    @Override
+    public void update(Coach coach) {
+        if (coach == null) {
+            throw new IllegalArgumentException("Coach cannot be null");
+        }
+        entityManager.merge(coach);
+    }
+
+    @Override
+    public void delete(Coach coach) {
+        if (coach == null) {
+            throw new IllegalArgumentException("Coach cannot be null");
+        }
+        entityManager.remove(entityManager.merge(coach));
+    }
+
+    @Override
+    public Coach getCoachById(Long id) {
+        return entityManager.find(Coach.class, id);
+    }
+
+    @Override
+    public List<Coach> getAllCoaches() {
+        return entityManager.createQuery("SELECT c FROM Coach c", Coach.class).getResultList();
+    }
+
+    @Override
+    public Coach getCoachByEmail(String email) {
+        return entityManager
+                .createQuery("SELECT c FROM Coach c where c.email = :mail", Coach.class)
+                .setParameter("mail", email)
```

Comment:
why not call the parameter `:email` then? Would it clash with something? 🤔

Reply:
OK, I get your point, though "email" and "mail" have different meanings in English. If you wish to differentiate between the variables, I'd suggest a tack different way. E.g. `comparisonEmail` or `providedEmail` or such. But I think most programmers prefer conciseness before verboseness and here all the `email` identifiers would have clear meaning. Namely `c.email`, `:email`, `String email`